lightning/api_grpc.go

```diff
@@ -218,3 +225,509 @@ func (l *LndGrpcLightningAPI) GetChanInfo(ctx context.Context, chanID uint64) (*
+	// We will first try obtaining ForwadingHistory and then move to SubscribeHtlc
+	const maxErrors = 5
+	sleepTime := 5 * time.Second
+
+	errorChan := make(chan ErrorData, 1)
+	outChan := make(chan []ForwardingEvent)
+
+	if batchSize == 0 {
+		batchSize = 50
+	}
+
+	errors := 0
+	req := &lnrpc.ForwardingHistoryRequest{
+		NumMaxEvents: uint32(batchSize),
+		IndexOffset:  uint32(0),
+		StartTime:    uint64(since.Unix()),
+	}
+
+	go func() {
+		var (
+			subscribeClient routerrpc.Router_SubscribeHtlcEventsClient
+		)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				// Do nothing
+			}
+			resp, err := l.Client.ForwardingHistory(ctx, req)
+
+			if err != nil {
+				glog.Warningf("Error getting ForwadingHistory %v\n", err)
+				if errors >= maxErrors {
+					errorChan <- ErrorData{Error: err, IsStillRunning: false}
+					return
+				}
+				errorChan <- ErrorData{Error: err, IsStillRunning: true}
+
+				time.Sleep(sleepTime)
+				errors++
+				continue
+			}
+
+			if resp.ForwardingEvents == nil || len(resp.ForwardingEvents) == 0 {
+				break
+			}
+
+			batch := make([]ForwardingEvent, 0, batchSize)
+
+			for _, event := range resp.ForwardingEvents {
+				batch = append(batch, ForwardingEvent{
+					Timestamp:     time.Unix(0, int64(event.TimestampNs)),
+					ChanIDIn:      event.ChanIdIn,
+					ChanIDOut:     event.ChanIdOut,
+					AmountInMsat:  event.AmtInMsat,
+					AmountOutMsat: event.AmtOutMsat,
+					FeeMsat:       event.FeeMsat,
+					IsSuccess:     true,
+					FailureString: "",
+				})
+			}
+
+			outChan <- batch
+			req.IndexOffset = resp.LastOffsetIndex
+
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+					// Do nothing
+				}
+
+				subscribeClient, err = l.RouterClient.SubscribeHtlcEvents(ctx, &routerrpc.SubscribeHtlcEventsRequest{})
+
+				if err != nil {
+					glog.Warningf("Error calling SubscribeHtlcEvents %v\n", err)
+					if errors >= maxErrors {
+						errorChan <- ErrorData{Error: err, IsStillRunning: false}
+						return
+					}
+					errorChan <- ErrorData{Error: err, IsStillRunning: true}
+
+					time.Sleep(sleepTime)
+					errors++
+					continue
+				} else {
+					break
+				}
+			}
+
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+					// Do nothing
+				}
+
+				event, err := subscribeClient.Recv()
+				if err == io.EOF {
+					errorChan <- ErrorData{Error: err, IsStillRunning: false}
+					return
+				}
+
+				if err != nil {
+					glog.Warningf("Error getting HTLC data %v\n", err)
+					if errors >= maxErrors {
+						errorChan <- ErrorData{Error: err, IsStillRunning: false}
+						return
+					}
+					errorChan <- ErrorData{Error: err, IsStillRunning: true}
+
+					time.Sleep(sleepTime)
+					errors++
+					continue
+				}
+
+				if event.EventType != routerrpc.HtlcEvent_FORWARD {
+					// Ignore non-forward events
+					continue
+				}
+
```

Comment:
this go routine is over complex. please split logic in sane size. Maybe try to write tests for it first and you will see its too complex.